OMCompiler/Compiler/NFFrontEnd/NFExpression.mo

```diff
@@ -5699,16 +5701,52 @@ public
         then
           SUBSCRIPTED_EXP(outExp, recordExp.subscripts, ty, recordExp.split);
 
+      case IF()
+        algorithm
+          trueBranch  := nthRecordElement(index, recordExp.trueBranch);
+          falseBranch := nthRecordElement(index, recordExp.falseBranch);
+        then
+          IF(typeOf(trueBranch), recordExp.condition, trueBranch, falseBranch);
+
       else
         algorithm
           Type.COMPLEX(cls = node) := typeOf(recordExp);
           node := Class.nthComponent(index, InstNode.getClass(node));
+          outExp := RECORD_ELEMENT(recordExp, index, InstNode.name(node), InstNode.getType(node));
         then
-          RECORD_ELEMENT(recordExp, index, InstNode.name(node), InstNode.getType(node));
+          inlineRecordConstructorElements(outExp);
 
     end match;
   end nthRecordElement;
 
+  function inlineRecordConstructorElements
+    "removes indexed constructor element calls
+    Constructor(a,b,c)[2] --> b"
+    input output Expression exp;
+  algorithm
+    exp := match exp
+      local
+        Expression new_exp;
+        Call call;
+        Function.Function fn;
+
+      case RECORD_ELEMENT(recordExp = CALL(call = call as Call.TYPED_CALL(fn = fn))) algorithm
+        if Function.Function.isDefaultRecordConstructor(fn) then
+          new_exp := listGet(call.arguments, exp.index);
+        elseif Function.Function.isNonDefaultRecordConstructor(fn) then
+          // ToDo: this has to be mapped correctly with the body.
+          //   for non default record constructors its not always the
+          //   case that inputs map 1:1 to attributes
+          new_exp := listGet(call.arguments, exp.index);
```

Comment:
This isn't true for default constructors either, since `final` components will not be part of the constructor. I.e.:
```Modelica
record R
  final parameter Real x = 1.0;
  Real y;
end R;
```
results in a record constructor:
```Modelica
function R
  protected Real x = 1.0;
  input Real y;
  output R res;
end R;
```
This might cause some issues since it means `nthRecordElement` with index = 1 will now return the value of `x` if the expression is a record expression but `y` if it's a default constructor call.

Reply:
True, we put no thought into those more complicated cases. I guess in the backend it didn't do as much harm as it will in the frontend. In the long run we should fix this, but for now I'll just put it back where it was...

Reply:
On second thought I'm not sure if it would actually cause any issues in the frontend since it converts all constructor calls to record expressions when typing them in `Call.typeCall`. So unless I'm forgetting something I don't think we'd hit that case in the frontend. But it's of course not a good idea to rely on that.